src/main/java/de/qabel/desktop/storage/AbstractNavigation.java

```diff
@@ -0,0 +1,259 @@
+		} catch (QblStorageNotFound e) {
+			logger.info("Could not reload metadata");
+		}
+		// the remote version has changed from the _old_ version
+		if ((updatedDM != null) && (!Arrays.equals(version, updatedDM.getVersion()))) {
+			logger.info("Conflicting version");
+			// ignore our local directory metadata
+			// all changes that are not inserted in the new dm are _lost_!
+			dm = updatedDM;
+			for (FileUpdate update: updatedFiles) {
+				handleConflict(update);
+			}
+			dm.commit();
+		}
+		uploadDirectoryMetadata();
+		for (String ref: deleteQueue) {
+			writeBackend.delete(ref);
+		}
+		// TODO: make a test fail without these
+		deleteQueue.clear();
+		updatedFiles.clear();
+	}
+
+	private void handleConflict(FileUpdate update) throws QblStorageException {
+		BoxFile local = update.updated;
+		BoxFile newFile = dm.getFile(local.name);
+		if (newFile == null) {
+			try {
+				dm.insertFile(local);
+			} catch (QblStorageNameConflict e) {
+				// name clash with a folder or external
+				local.name = conflictName(local);
+				// try again until we get no name clash
+				handleConflict(update);
+			}
+		} else if (newFile.equals(update.old)) {
+			logger.info("No conflict for the file " + local.name);
+		} else {
+			logger.info("Inserting conflict marked file");
+			local.name = conflictName(local);
+			if (update.old != null) {
+				dm.deleteFile(update.old);
+			}
+			if (dm.getFile(local.name) == null) {
+				dm.insertFile(local);
+			}
+		}
+	}
+
+	private String conflictName(BoxFile local) {
+		return local.name + "_conflict_" + local.mtime.toString();
+	}
+
+	protected abstract void uploadDirectoryMetadata() throws QblStorageException;
+
+	@Override
+	public BoxNavigation navigate(BoxExternal target) {
+		throw new NotImplementedException("Externals are not yet implemented!");
+	}
+
+	@Override
+	public List<BoxFile> listFiles() throws QblStorageException {
+		return dm.listFiles();
+	}
+
+	@Override
+	public List<BoxFolder> listFolders() throws QblStorageException {
+		return dm.listFolders();
+	}
+
+	@Override
+	public List<BoxExternal> listExternals() throws QblStorageException {
+		//return dm.listExternals();
+		throw new NotImplementedException("Externals are not yet implemented!");
+	}
+
+	@Override
+	public BoxFile upload(String name, File file) throws QblStorageException {
+		KeyParameter key = cryptoUtils.generateSymmetricKey();
+		String block = UUID.randomUUID().toString();
+		BoxFile boxFile = new BoxFile(block, name, file.length(), 0L, key.getKey());
+		boxFile.mtime = uploadEncrypted(file, key, "blocks/" + block);
+		// Overwrite = delete old file, upload new file
+		BoxFile oldFile = dm.getFile(name);
+		if (oldFile != null) {
+			deleteQueue.add(oldFile.block);
+			dm.deleteFile(oldFile);
+		}
+		updatedFiles.add(new FileUpdate(oldFile, boxFile));
+		dm.insertFile(boxFile);
+		return boxFile;
+	}
+
+	protected long uploadEncrypted(File file, KeyParameter key, String block) throws QblStorageException {
```

Comment:
Doesn't help if the subclasses are not thrown. In this example the ```InvalidKeyException``` which could be very helpful is hidden behind the ```QblStorageException``` (completely inspecific). Something like a ```QblDecryptionFailedException extends QblStorageException``` or ```QblInvalidKeyException extends QblStorageException``` could be handled in a useful way (like showing some faq-text to the user or opening a key import screen ...)